src/lib.rs

```diff
@@ -346,10 +345,425 @@ impl From<Error> for io::Error {
+    /// [`std::path::Path::read_link()`]: https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.read_link
+    fn read_link(&self) -> Result<PathBuf> {
+        Path::read_link(self.borrow())
+            .map_err(|err| Error::new(
+                err,
+                "reading link target of",
+                self.clone().into(),
+            ))
+    }
+
+    /// Returns the canonical, absolute form of the path with all intermediate
+    /// components normalized and symbolic links resolved.
+    ///
+    /// The same as [`std::path::Path::canonicalize()`], except:
+    ///
+    ///   - On success, returns a `path_abs::PathAbs` instead of a `PathBuf`
+    ///   - returns a rich [`path_abs::Error`] when a problem is encountered
+    ///
+    /// [`path_abs::Error`]: struct.Error.html
+    /// [`std::path::Path::canonicalize()`]: https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.canonicalize
+    fn canonicalize(&self) -> Result<PathAbs> {
+        Path::canonicalize(self.borrow())
+            .map(|path| PathAbs(path.into()))
+            .map_err(|err| Error::new(
+                err,
+                "canonicalizing",
+                self.clone().into(),
+            ))
+    }
+
+    /// Returns the path without its final component, if there is one.
+    ///
+    /// The same as [`std::path::Path::parent()`], except that it returns a
+    /// `Result` with a rich [`path_abs::Error`] when a problem is encountered.
+    ///
+    /// [`path_abs::Error`]: struct.Error.html
+    /// [`std::path::Path::parent()`]: https://doc.rust-lang.org/stable/std/path/struct.Path.html#method.parent
+    fn parent(&self) -> Result<&Path> {
+        let parent_path = Path::parent(self.borrow());
+        if let Some(p) = parent_path {
+            Ok(p)
+        } else {
+            Err(Error::new(
+                io::Error::new(io::ErrorKind::NotFound, "path has no parent"),
+                "truncating to parent",
+                self.clone().into(),
+            ))
+        }
+    }
+}
+
+impl<T> PathInfo for T where T: Clone + Borrow<PathBuf> + Into<Arc<PathBuf>> {}
+
+/// Methods that modify a path.
+///
+/// These methods are not implemented for all `path_abs` types because they
+/// may break the type's invariant. For example, if you could call
+/// `truncate_to_parent()` on a `PathFile`, it would no longer be the path to
+/// a file, but the path to a directory.
+///
+/// As a general rule, methods that can return an error will return a rich
+/// [`path_abs::Error`] instead of a [`std::io::Error`].
+pub trait PathMut: PathInfo {
+    /// Appends `path` to this path.
+    ///
+    /// Note that this method represents pure concatenation, not "adjoinment"
+    /// like [`PathBuf::push`].
+    ///
+    /// # Errors
+    ///
+    /// This method returns an error if the `path` parameter contains a prefix
+    /// component (on Windows, like `C:`), or enough `..` components to consume
```

Comment:
I don't think this is true for `PathBuf` right?

Reply:
I'm not sure what you mean. `PathBuf` doesn't have an `append()` method at all.

Reply:
Are we not [`impl PathMut for PathBuf`](https://github.com/vitiral/path_abs/pull/36/files#diff-b4aea3e418ccdb71239b96952d9cddb6R587)?

Reply:
Sure, but we can `impl` it to do whatever we want. Right now, the `impl PathMut for PathBuf` behaves (as best I can tell) as the docstring describes, and every other `impl` chains back to that one.

Reply:
I would expect that you can have a relative path for `PathBuf`, but it states that it will return an error if:

```
enough `..` components to consume this path
```

Reply:
There's a test for that exact behaviour on lines 970 (for Windows) and 1131 (for POSIX).